src/workerd/api/messagechannel.c++

```diff
@@ -0,0 +1,317 @@
+    return MessageEvent::NewError(
+        js, jsg::JsRef(js, jsg::JsValue(exception.getHandle(js))), kj::str(), port->addRef(), {});
+  });
+
+  // If any of the message/messageerror event handlers throw,
+  // capture the error and pass it to reportError instead of
+  // propagating up.
+  js.tryCatch([&] { port->dispatchEventImpl(js, kj::mv(event)); }, [&](jsg::Value exception) {
+    auto context = js.v8Context();
+    auto& global =
+        jsg::extractInternalPointer<ServiceWorkerGlobalScope, true>(context, context->Global());
+    global.reportError(js, jsg::JsValue(exception.getHandle(js)));
+  });
+}
+
+// Deliver the message to all the jsrpc remotes we have
+kj::Promise<void> MessagePort::sendToRpc(kj::Own<rpc::JsValue::Reader> message) {
+  KJ_IF_SOME(outputLocks, ioContext.waitForOutputLocksIfNecessary()) {
+    co_await outputLocks;
+  }
+  kj::Vector<kj::Promise<void>> promises;
+  for (rpc::JsMessagePort::Client cap: rpcClients) {
+    auto req = cap.callRequest();
+    req.setData(*message);
+    // TODO(message-port): Removing the port when dropped?
+    promises.add(req.send().ignoreResult());
+  }
+  co_await kj::joinPromises(promises.releaseAsArray());
+}
+
+// Deliver the message to this port, buffering if necessary if the port
+// has not been started. Buffered messages will be delivered when the
+// port is started later.
+void MessagePort::deliverMessage(jsg::Lock& js, rpc::JsValue::Reader value) {
+  KJ_SWITCH_ONEOF(state) {
+    KJ_CASE_ONEOF(pending, Pending) {
+      // We have not yet started the port so buffer to message.
+      pending.add(capnp::clone(value));
+    }
+    KJ_CASE_ONEOF(started, Started) {
+      ioContext.addTask(sendToRpc(capnp::clone(value)));
+      deliver(js, capnp::clone(value), addRef());
+    }
+    KJ_CASE_ONEOF(_, Closed) {
+      // Nothing to do in this case. Drop the message on the floor.
+    }
+  }
+}
+
+// Binds two ports to each other such that messages posted to one
+// are delivered on the other.
+void MessagePort::entangle(MessagePort& port1, MessagePort& port2) {
+  port1.other = port2.addRef();
+  port2.other = port1.addRef();
+}
+
+// Post a message to the entangled port.
+void MessagePort::postMessage(jsg::Lock& js,
+    jsg::Optional<jsg::JsRef<jsg::JsValue>> data,
+    jsg::Optional<TransferListOrOptions> options) {
+
+  KJ_IF_SOME(opt, options) {
+    // We don't currently support transfer lists, even for local
+    // same-isolate delivery.
+    // TODO(conform): Implement transfer later?
+    KJ_SWITCH_ONEOF(opt) {
+      KJ_CASE_ONEOF(list, TransferList) {
+        JSG_REQUIRE(list.size() == 0, Error, "Transfer list is not supported");
+      }
+      KJ_CASE_ONEOF(opts, PostMessageOptions) {
+        KJ_IF_SOME(list, opts.transfer) {
+          JSG_REQUIRE(list.size() == 0, Error, "Transfer list is not supported");
+        }
+      }
+    }
+  }
+
+  KJ_IF_SOME(o, other) {
+    // TODO(message-port): Set up the external handler to support more types.
```

Comment:
This is currently not set up to allow for other types we would send over jsrpc but it can be. Should it be? What if there are no jsrpc targets and the `MessagePort` is just cloning locally? Should we still use the jsrpc external handler in that case? It's not clear... so initially this will only take regularly serializable objects.